### PR TITLE
upgrade deps: upgrade monix / cats-effect etc.

### PR DESCRIPTION
**Pull request by ahjohannessen**
- upgrade monix to cats-effect 1.0.0 compatible version
   and fix deprecated usage of `concat` and `merge` in
   monix 3.

 - fix tests in `LifecycleHookSpec` that broke as a result
   moving from monix RC1 to RC2. It is seems related to some
   acynchronicity issue wrt. future usage. Alexandru from
   monix said that there have been changes in
   `TrampolineExecutionContext` that might have introduced a
   bug.

   Moreover, `onNext` can be synchronous if the listeners are
   synchronous — this is an implementation detail and is not
   guaranteed.

 - bump scala / scalajs

### CODE REVIEW

File: src/test/scala/outwatch/LifecycleHookSpec.scala [hunk 1]
```diff
@@ -132,20 +132,15 @@ class LifecycleHookSpec extends JSDomAsyncSpec {
       node <- div(message, onSnabbdomUpdate --> sink1)(onSnabbdomUpdate --> sink2)
     } yield node
 
-    val res1 = for {
-      _ <- renderUnsafe(node)
-    } yield {
-      switch1 shouldBe false
-      switch2 shouldBe false
-    }
+    renderUnsafe(node).value.get.get
```
[the remark this answers is not in the document]
Comment (ahjohannessen) [reply src/test/scala/outwatch/LifecycleHookSpec.scala]: It is needed to await the future, using this imperative style. Using flatmap somehow does not work with `onNext`

[the remark this answers is not in the document]
Comment (fdietze) [reply src/test/scala/outwatch/LifecycleHookSpec.scala]: I understand now, that this would throw if the future was not executed successfully or not executed at all. But why doesn't it work in the for-loop?

[the remark this answers is not in the document]
Comment (ahjohannessen) [reply src/test/scala/outwatch/LifecycleHookSpec.scala]: Please try it out if you have the time, on the phone now.

[the remark this answers is not in the document]
Comment (ahjohannessen) [reply src/test/scala/outwatch/LifecycleHookSpec.scala]: Some kind of race condition, I guess. It is the equivalent of unsafeRunSync

[the remark this answers is not in the document]
Comment (fdietze) [reply src/test/scala/outwatch/LifecycleHookSpec.scala]: > It is needed to await the future
`.value` actually does not await. It just takes the current value...
(see https://www.scala-lang.org/api/current/scala/concurrent/Future.html#value:Option[scala.util.Try[T]])

I'm investigating in the for-loop problem now.

[the remark this answers is not in the document]
Comment (ahjohannessen) [reply src/test/scala/outwatch/LifecycleHookSpec.scala]: Ok thanks

[the remark this answers is not in the document]
Comment (fdietze) [reply src/test/scala/outwatch/LifecycleHookSpec.scala]: It looks like we cannot rely on the backpressure of `message.onNext` to give us a `Continue` *after* `onSnabbdomUpdate` was called and its emitter has written into the sink. There seems to be some asyncronicity in there somewhere. I suspect it to be the Emitter. This is especially strange since we have a synchronous scheduler...

Still, thank you very much for taking the time and make this PR.

Maybe @cornerman has an idea why we cannot rely on the backpressure?

[the remark this answers is not in the document]
Comment (ahjohannessen) [reply src/test/scala/outwatch/LifecycleHookSpec.scala]: @fdietze I suspected it was something around `onNext`

[the remark this answers is not in the document]
Comment (ahjohannessen) [reply src/test/scala/outwatch/LifecycleHookSpec.scala]: @fdietze Alexandru from Monix says nothing has changed wrt. backpressure/scheduler behavior between RC1 and RC2.

**Comment by cornerman** [comment]
@ahjohannessen interesting! thank you very much for investigating and asking in the monix channel! Should we wait for a potential fix on the monix side?

File: yarn.lock [hunk 2]
```diff
@@ -2,3367 +2,526 @@
 # yarn lockfile v1
```
Comment (cornerman) [inline yarn.lock]: please revert the changes in the yarn.lock. I am not sure why these things change and how we should handle changes with different node/yarn versions.

**Comment by ahjohannessen** [comment]
Perhaps wait a bit, Alexandru said he would investigate.

File: yarn.lock (hunk 2 above)
Comment (ahjohannessen) [reply yarn.lock]: Will do

**Comment by ahjohannessen** [comment]
@cornerman
>Should we wait for a potential fix on the monix side?

Perhaps not, as I have changed tests to be defined in terms of `IO` and all of them are converted to `Future` in the last possible momemt, because there exists an implicit converstion in `OutwatchSpec`:

```scala
implicit def ioAssertionToFutureAssertion(io: IO[Assertion]): Future[Assertion] = 
  io.unsafeToFuture()
```

Moreover, `onNext` can be synchronous if the listeners are synchronous — this is an implementation detail and is not guaranteed. Also, when using

```scala
def renderUnsafe(node: dom.VNode): Future[Unit] = 
  OutWatch.renderInto("#app", node).unsafeToFuture()
```

`renderUnsafe(node).map(???)` might trigger an async boundary that messes things up.

//cc @fdietze @LukaJCB

**Comment by cornerman** [comment]
Thanks for the update!

I think, we should wait until the issue was investigated in monix. We updated monix and the code from before does not work anymore, so I would like to wait and see what happened there.

Furthermore, I am not sure whether adding more conversion methods like `toIO` is so great. It kind of makes the tests more complex, as we now have to convert every future into an IO when it is used in combination.

**Comment by ahjohannessen** [comment]
@cornerman I asked Alexandru earlier, he hadn’t had time to look into it yet, but hopefully today.

**Comment by alexandru** [comment]
Hi guys,

So I took a look at `TrampolineScheduler`, my initial suspicion and couldn't find anything wrong with it. That trampoline implementation changed a little, but it's used everywhere in Monix and it can't be the reason.

TL;DR — I don't know why the test fails, but lets discuss it.

@ahjohannessen on what we've discussed on Gitter, you gave me this sample:

```scala
    renderUnsafe(node).map { _ ⇒
      switch1 shouldBe false
      switch2 shouldBe false
      // This prints "1":
      println(message.size)
      message.onNext("next")
      // Fails:
      switch1 shouldBe true
      switch2 shouldBe true
    }
```

First of all, it's clear that the reason this fails is because `message.onNext("next")` introduces an async boundary. And by introducing an async boundary, in this case it doesn't matter if the scheduler is a `TrampolineScheduler`, because the operation is scheduled after `renderUnsafe` is being executed, which means that it is delayed.

So for example when you do this:

```scala
renderUnsafe(node).value.get.get
switch1 shouldBe false
switch2 shouldBe false

message.onNext("next").value.get.get
switch1 shouldBe true
switch2 shouldBe true
```

This works because you're forcing the `TrampolineScheduler` to execute `renderUnsafe` and then you're forcing to execute `message.onNext` before the final comparisons. To give an example, consider this:

```scala
    implicit val ec = TrampolineScheduler(Scheduler.global, SynchronousExecution)

    val f1 = Future(1)
    val p2 = Promise[Int]()

    f1.map { _ =>
      // Scheduling another future for execution
      val f2 = Future(2)
      p2.completeWith(f2)
      // This will be None: <--- Witness this!
      println(f2.value)
    }

    // This will be completed ;-)
    println(p2.future.value)
```

Witness that `f2.value`, it is going to be `None`, even if we are using the trampoline, which is supposed to execute everything on the same thread. The problem is, the execution may be on the same thread, but it's not immediate, it gets delayed and it behaves like so in order to avoid stack overflow errors.

So to get one thing out of the way

1. `PublishSubject` will return a completed `Continue` only if all of its subscribers return a synchronous `Continue` and this is only an internal optimization you can't rely on long term, it's not guaranteed so to speak
2. in that sample with `println(message.size)` you said it printed `1`, which means I don't have complete information — clearly we aren't talking about the 2 synchronous `Sync` objects you're creating, those are not subscribing directly to your `PublishSubject`, but indirectly (please correct me if I'm wrong)
3. if that subscriber that subscribes to `PublishSubject` returns a `Future` that's not a straight, synchronous `Continue`, an asynchronous boundary is inserted via `message.onNext` and hence the effect we see here

Unfortunately this doesn't answer why your test is failing in the first place.

We can say however that we're dealing with an extra async boundary that wasn't there before, messing up the ordering somehow. I'm suspecting something like this:

```scala
val thirdParty = PublishSubject[String]()
thirdParty.subscribe(sink1)
thirdParty.subscribe(sink2)

message.subscribe { elem =>
  // Extra async boundary
  Future(thirdParty.onNext(elem))
  // No back-pressure
  Continue
}
```

Could you think of anywhere in your test of where this could happen?

**Comment by alexandru** [comment]
Btw, I tried taking a look at the code that was failing, I see now that Travis is passing, I assume it is because you changed to usage of `IO`.

Changing to `IO` via `IO.fromFuture` does nothing but to avoid the `TrampolineScheduler`. And assuming the hypothesis that we've got an extra async boundary somewhere, this makes sense, because by avoiding the `TrampolineScheduler`, it somehow corrects the ordering.

This isn't a good thing though. The problem is still there and we'll have to find it.

Can you create a branch for me with the tests failing, so I can compile and run the code myself, sometime tomorrow?

**Comment by cornerman** [comment]
I updated the PR #247 to monix-RC2 as well. It works there. Though I do not wait on the backpressure futures in the test there.

**Comment by alexandru** [comment]
Guys, sorry for taking so long to investigate.

I think I figured it out — `TrampolineScheduler` changed internally to use a Stack instead of a Queue.

What this does is to reverse the order of execution in certain situations, so if you have `task1` and `task2` registered in this order, you're going to have `task2` execute before `task1` with a stack.

This isn't necessarily OK on Monix's side, we should have a queue in there and I'm changing the implementation as we speak. 

However what this also means is that your test fails for good reasons and it's not Monix.

After I do the change in `TrampolineScheduler`, I'll publish another hash version immediately to validate that what I'm saying is true.

**Comment by cornerman** [comment]
@ahjohannessen I think for updating monix and cats-effect, the other branch already takes care of this. So now this PR here is a bit redundant. If you want to work further on the tests, it would be great if you could rebase those changes on that other branch. When merging the new tests in LifecycleHookSpec into that branch, I could not keep all the changes because it was just too much to merge.

**Comment by ahjohannessen** [comment]
Thanks for the heads up @alexandru :)

**Comment by cornerman** [comment]
@alexandru thank for looking into it! Great that it will be fixed.

> However what this also means is that your test fails for good reasons and it's not Monix.
I am not sure what this actually means. What kind of construction could be a reason for this behaviour? Do you have a suggestion what we could improve?

**Comment by alexandru** [comment]
@cornerman let me prove it by changing the implementation first.

My assumption is that the ordering of those two operations (e.g. `renderUnsafe` and `message.onNext`) is actually reversed when `TrampolineScheduler` uses a `Stack` underneath instead of a `Queue`, which can either mean two things:

1. your `renderUnsafe` actually finishes after `message.onNext`, or
2. `message.onNext` hits some sort of async buffer, meaning that it can finish before the message is received by your 2 sinks

Or in other words, if the test worked before, it was due to implementation details that aren't guaranteed. 

This is why I prefer `TestScheduler` when testing async code, whenever possible, because `TestScheduler` randomizes the execution in order to simulate non-determinism.`TestScheduler` doesn't work for actual asynchrony, it just simulates async execution, so you can't do actual I/O with it. 

Before doing any changes, let me make the switch back to usage of a Queue first. Because if the test works afterwards, then I'm right about the 2 possible causes I listed above. And note this doesn't necessarily imply a bug in your code. Maybe it's just the test.

**Comment by alexandru** [comment]
Published a test version at: `3.0.0-RC2-379831b`

I can confirm that the `outwatch.LifecycleHookSpec` tests are passing after switching to it.

So I was right, it's due to switching from a queue to a stack, which reorders the execution of registered `Runnable`s (which often correspond to `Future`s). I do suggest finding out why this happens and if possible switch to a `TestScheduler`, because it would make the tests better.

**Comment by alexandru** [comment]
Btw, changing the trampoline implementation to a queue is permanent, so no need to worry about it going forward: https://github.com/monix/monix/pull/765

**Comment by alexandru** [comment]
@cornerman yes, I think that’s the case.

I don’t exactly know the reason, but one of those two future-enabled operations wasn’t waiting on the result as it should and a simple change of the scheduler internals exposed it. 

For the tests where you can afford TestScheduler, definitely use it.

**Comment by cornerman** [comment]
Great, will try it out :)

**Comment by alexandru** [comment]
Btw, I released 3.0.0-RC2 today. It's basically the same as that last hash version, but you might want to switch to it.

Final 3.0.0 is scheduled in 2 weeks from now.

Cheers,

**Comment by cornerman** [comment]
Awesome! Thank you.

**Comment by cornerman** [comment]
@ahjohannessen 
> Are my changes to LifecycleHookSpec necessary in your branch?

A lot of sideeffects in LifecycleHookSpec are gone due to the changes, so less changes are needed. Still, your refactoring adhered to the backpressure protocol and I do not. So, it might be good for switching to the TestScheduler (see #249) and for checking if we can still reproduce the problem we had before.

**Comment by cornerman** [comment]
I think, I found the problem with the improper backpressure. When working on some related tests, I suddenly got a similar issue. The root cause was using two different execution contexts in the test: One being the `Scheduler` and one being the `ExecutionContext` provided by scalatest async suite. I have fixed it in this PR: https://github.com/OutWatch/outwatch/pull/254/commits/2d7ae5d36af129c89c8fe6bb56b1d899a1cc04c0

**Comment by ahjohannessen** [comment]
@cornerman Oh wow, great catch if that is the root cause 👍

**Comment by cornerman** [comment]
The update is now on master. So I am closing this one, please reopen if there is something missing still.

Thank you all for your work and investigation here!

**Comment by ahjohannessen** [comment]
:+1:
